### PR TITLE
Mafia: allow JSON game setups to be used instead of lists of roles

### PR DESCRIPTION
**Pull request by bumbadadabum**
I'll add the ability to export game setups as well at a later date.

### CODE REVIEW

File: chat-plugins/mafia.js [hunk 1]
```diff
@@ -577,23 +580,64 @@ exports.commands = {
...
+				targetObj = JSON.parse(target);
+			} catch (e) {}
+
+			if (targetObj) {
+				if (!(targetObj.classes && Object.keys(targetObj.classes).length)) return this.errorReply("Invalid input.");
+
+				let roleList = [];
+
+				for (let i in targetObj.classes) {
+					if (!(i in MafiaData.MafiaClasses)) {
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Guard for inherited properties

File: chat-plugins/mafia.js [hunk 2]
```diff
@@ -577,23 +580,64 @@ exports.commands = {
...
+						return this.errorReply(Tools.escapeHTML(i) + " is not a valid mafia class.");
+					}
+
+					let amt = parseInt(targetObj.classes[i]);
+					if (isNaN(amt)) return this.errorReply("Invalid amount for class " + Tools.escapeHTML(i));
```
Comment (Slayer95) [inline chat-plugins/mafia.js]: Guard for negative/unsafe quantities

**Comment by bumbadadabum** [comment]
Nicely spotted, @Slayer95. I didn't think of inherited properties myself at all.

**Comment by Zarel** [comment]
You need to heavily sanitize JSON in general. I think we have a `getString` command for safely extracting strings from JSON.

**Comment by bumbadadabum** [comment]
This is fine.

**Pull request merged**